azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/tools/auth/AzureAuthManager.java

```diff
@@ -0,0 +1,101 @@
+import com.microsoft.azure.tools.auth.model.AuthConfiguration;
+import com.microsoft.azure.tools.auth.model.AuthType;
+import com.microsoft.azure.tools.auth.model.AzureCredentialWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Single;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class AzureAuthManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AzureAuthManager.class);
+    private static final String INVALID_AZURE_ENVIRONMENT = "Invalid environment string '%s', please replace it with one of " +
+            "\"AzureCloud\",  \"AzureChinaCloud\",  \"AzureUSGovernment\",  \"AzureGermanCloud\".";
+    private static final String FALLBACK_TO_AUTO_AUTH_TYPE = "Will use 'auto' by default.";
+
+    public static Single<AzureCredentialWrapper> getAzureCredentialWrapper(AuthConfiguration configuration) {
+        AuthConfiguration auth = configuration == null ? new AuthConfiguration() : configuration;
+        AzureEnvironment env;
+        try {
+            env = getAzureEnvironmentFromString(auth.getEnvironment());
+        } catch (InvalidConfigurationException e) {
+            return Single.error(e);
+        }
+        ChainedCredentialRetriever retrievers = new ChainedCredentialRetriever();
+        AuthType authType = getAuthTypeFromString(auth.getType());
+        Map<AuthType, ICredentialRetriever> retrieverMap = buildCredentialRetrievers(env);
+        if (authType.equals(AuthType.AUTO)) {
+                retrievers.addRetriever(new ServicePrincipalCredentialRetriever(auth, env));
+                retrievers.addRetriever(retrieverMap.get(AuthType.MANAGED_IDENTITY));
+                retrievers.addRetriever(retrieverMap.get(AuthType.AZURE_CLI));
+                retrievers.addRetriever(retrieverMap.get(AuthType.VSCODE));
+                retrievers.addRetriever(retrieverMap.get(AuthType.VISUAL_STUDIO));
+                retrievers.addRetriever(retrieverMap.get(AuthType.OAUTH2));
+                retrievers.addRetriever(retrieverMap.get(AuthType.DEVICE_CODE));
+        } else {
+            // for specific auth type:
+            if (AuthType.SERVICE_PRINCIPAL == authType) {
+                retrievers.addRetriever(new ServicePrincipalCredentialRetriever(auth, env));
+            } else {
+                if (!retrieverMap.containsKey(authType)) {
+                    return Single.error(new UnsupportedOperationException(String.format("authType '%s' not supported.", authType)));
+                }
+                retrievers.addRetriever(retrieverMap.get(authType));
+            }
+        }
+        return retrievers.retrieve().onErrorResumeNext(e ->
+                Single.error(new LoginFailureException(String.format("Cannot get credentials from authType '%s' due to error: %s", authType, e.getMessage())))
+        );
+    }
+
+    private static Map<AuthType, ICredentialRetriever> buildCredentialRetrievers(AzureEnvironment env) {
```

Comment:
Make the method as a static map?

Reply:
May use linked map here which could be used in auto mode as well

Reply:
we need env as parameter

Reply:
May use `LinkedMap<AuthType, Function<Environment, Retriever>>`? As `buildCredentialRetrievers` will create all `Retriever` each time when invoked